oauth2client/contrib/multiprocess_file_storage.py

```diff
@@ -0,0 +1,261 @@
+
+    def __init__(self, filename):
+        self._file = None
+        self._filename = filename
+        self._process_lock = fasteners.InterProcessLock(
+            '{}.lock'.format(filename))
+        self._thread_lock = threading.Lock()
+        self._read_only = False
+        self._credentials = {}
+
+    def _create_file_if_needed(self):
+        """Creates the an empty credential storage file if it does not
+        exist."""
+        if self._read_only:
+            return False
+
+        if not os.path.exists(self._filename):
+            open(self._filename, 'a+b').close()
+            logging.info('Credential file {} created'.format(self._filename))
+            return False
+
+        return True
+
+    def _load_credentials(self):
+        """(Re-)loads the credentials from the file."""
+        if not self._create_file_if_needed():
+            return
+
+        self._credentials.update(self._load_credentials_file(self._file))
+
+        logger.debug('Read credential file')
+
+    def _load_credentials_file(self, fh):
+        credentials = {}
+
+        try:
+            fh.seek(0)
+            data = json.load(fh)
+        except Exception:
+            logger.warning(
+                'Credentials file could not be loaded, will ignore and '
+                'overwrite.')
+            return credentials
+
+        if data.get('file_version') != 2:
+            data = {}
+            logger.warning(
+                'Credentials file is not version 2, will ignore and '
+                'overwrite.')
+            return credentials
+
+        for key, encoded_credential in iteritems(data.get('credentials', {})):
+            try:
+                credential_json = base64.b64decode(encoded_credential)
+                credential = Credentials.new_from_json(credential_json)
+                credentials[key] = credential
+            except:
+                logger.warning(
+                    'Invalid credential {} in file, ignoring.'.format(key))
+
+        return credentials
+
+    def _write_credentials(self):
+        if self._read_only:
+            logger.debug('In read-only mode, not writing credentials.')
+            return
+
+        self._create_file_if_needed()
+        self._write_credentials_file(self._file, self._credentials)
+        logger.debug('Wrote credential file {}.'.format(self._filename))
+
+    def _write_credentials_file(self, fh, credentials):
+        data = {'file_version': 2, 'credentials': {}}
+
+        for key, credential in iteritems(credentials):
+            credential_json = credential.to_json()
+            encoded_credential = base64.b64encode(credential_json)
+            data['credentials'][key] = encoded_credential
+
+        fh.seek(0)
+        json.dump(data, fh)
+        fh.truncate()
+
+    def acquire_lock(self):
+        self._thread_lock.acquire()
+        locked = self._process_lock.acquire(timeout=INTERPROCESS_LOCK_DEADLINE)
+
+        self._create_file_if_needed()
```

Comment:
If this can be done without holding the process lock, why wait until after attempting to acquire the process lock to do it? If this must be done while holding the process lock, why wait until after doing this to check whether or not the attempt to acquire the process lock succeeded?

Does this really belong in the limbo of "better to do while holding the process lock, but safe to do without the lock if that's the kind of day we're having"?